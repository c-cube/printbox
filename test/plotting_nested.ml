module B = PrintBox
module BPlot = PrintBox_ext_plot

(* This tests the printbox-ext-plot extension. *)

let reg_45 =
  B.(
    frame
      (vlist
         [
           text "123456789";
           frame ~stretch:true
             (vlist [ text "...." |> align ~h:`Right ~v:`Top; text "." ]);
         ]))

let for_3 =
  let n = 3 in
  Array.init n (fun i -> Array.init n (fun j -> B.sprintf "(%d,%d)" i j))
  |> B.grid

let nice_unicode =
  B.(
    frame
    @@ vlist
         [
           text "nice unicode! ğŸ’ª";
           frame
           @@ hlist
                [
                  vlist
                    [
                      text "oÃ¯ Ï‰Îµird nums:\nÏ€/2\nÏ„/4";
                      center_hv
                      @@ tree (text "0")
                           [ text "1"; tree (text "Ï‰") [ text "Ï‰Â²" ] ];
                    ];
                  frame @@ frame @@ frame
                  @@ vlist
                       [
                         text "sum=Î£_i aÂ·xáµ¢Â²\nâ€”â€”â€”â€”â€”\n1+1";
                         align_right @@ text "ÅŒâ‚’\nÃ€";
                       ];
                ];
         ])

let test ~size =
  B.extension
  @@ BPlot.(
       Plot
         {
           default_config with
           size;
           specs =
             [
               Scatterbag
                 {
                   points =
                     [|
                       (0., 1.), reg_45;
                       (1., 0.), for_3;
                       (0.75, 0.75), nice_unicode;
                     |];
                 };
               (* Map
                 {
                   callback =
                     (fun (x, y) ->
                       let s = ((x ** 2.) +. (y ** 2.)) ** 0.5 in
                       B.line
                       @@
                       if s < 0.3 then
                         " "
                       else if s < 0.6 then
                         "."
                       else if s < 0.9 then
                         ","
                       else if s < 1.2 then
                         ":"
                       else
                         ";");
                 }; *)
             ];
         })

let () =
  print_endline "Text output:";
  print_endline @@ PrintBox_text.to_string
  @@ test ~size:BPlot.default_config.size;
  print_endline "\nHTML output:";
  print_endline @@ PrintBox_html.to_string @@ test ~size:(800, 800)
