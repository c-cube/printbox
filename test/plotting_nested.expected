Text output:
┌──┬────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ 1│┌─────────┐                                                                                         │
│  ││123456789│                                                                                         │
│  │├─────────┤                                                                                         │
│  ││┌───────┐│                                                                                         │
│  │││   ....││                                                                                         │
│  ││├───────┤│                                                                                         │
│  │││.      ││                                                                                         │
│  ││└───────┘│                                                                                         │
│  │└─────────┘                                                                                         │
│  │                                                                                                    │
│  │                                                                          ┌─────────────────────────│
│  │                                                                          │nice unicode! 💪         │
│  │                                                                          ├─────────────────────────│
│  │                                                                          │┌──────────────┬─────────│
│  │                                                                          ││oï ωεird nums:│┌────────│
│  │                                                                          ││π/2           ││┌───────│
│  │                                                                          ││τ/4           │││┌──────│
│  │                                                                          │├──────────────┤│││sum=Σ_│
│  │                                                                          ││              ││││————— │
│y │                                                                          ││              ││││1+1   │
│  │                                                                          ││   0          │││├──────│
│  │                                                                          ││   ├─1        ││││      │
│  │                                                                          ││   └─ω        ││││      │
│  │                                                                          ││     └─ω²     │││└──────│
│  │                                                                          ││              ││└───────│
│  │                                                                          ││              │└────────│
│  │                                                                          │└──────────────┴─────────│
│  │                                                                          └─────────────────────────│
│  │                                                                                                    │
│  │                                                                                                    │
│  │                                                                                                    │
│  │                                                                                                    │
│  │                                                                                                    │
│  │                                                                                                    │
│  │                                                                                                    │
│  │                                                                                                    │
│  │                                                                                                    │
│  │                                                                                                    │
│  │                                                                                                    │
│ 0│                                                                                                   (│
├──┼────────────────────────────────────────────────────────────────────────────────────────────────────┤
│  │0                                                                                                  1│
│  │                                                 x                                                  │
└──┴────────────────────────────────────────────────────────────────────────────────────────────────────┘

HTML output:
<div><div style="border:thin solid"><table class="framed"><tr><td><table class="non-framed"><tr><td><div>y</div></td><td><table class="non-framed"><tr><td><div>1</div></td></tr><tr><td><div>0</div></td></tr></table></td></tr></table></td><td><div style="border:thin dotted;position:relative;width:800;height:800"><div style="position:absolute;top:0px;left:0px"><div style="border:thin solid"><table class="framed"><tr><td><div>123456789</div></td></tr><tr><td><div style="border:thin solid"><table class="framed"><tr><td><div class="align-right"><div>....</div></div></td></tr><tr><td><div>.</div></td></tr></table></div></td></tr></table></div></div><div style="position:absolute;top:200px;left:599px"><div style="border:thin solid"><table class="framed"><tr><td><div>nice unicode! 💪</div></td></tr><tr><td><div style="border:thin solid"><table class="framed"><tr><td><table class="framed"><tr><td><div>oï ωεird nums:<br/>π/2<br/>τ/4</div></td></tr><tr><td><div class="center"><div><div>0</div><ul><li><div>1</div></li><li><div><div>ω</div><ul><li><div>ω²</div></li></ul></div></li></ul></div></div></td></tr></table></td><td><div style="border:thin solid"><div style="border:thin solid"><div style="border:thin solid"><table class="framed"><tr><td><div>sum=Σ_i a·xᵢ²<br/>—————<br/>1+1</div></td></tr><tr><td><div class="align-right"><div>Ōₒ<br/>À</div></div></td></tr></table></div></div></div></td></tr></table></div></td></tr></table></div></div><div style="position:absolute;top:799px;left:799px"><table class="framed"><tr><td><div>(0,0)</div></td><td><div>(0,1)</div></td><td><div>(0,2)</div></td></tr><tr><td><div>(1,0)</div></td><td><div>(1,1)</div></td><td><div>(1,2)</div></td></tr><tr><td><div>(2,0)</div></td><td><div>(2,1)</div></td><td><div>(2,2)</div></td></tr></table></div></div></td></tr><tr><td><div></div></td><td><table class="non-framed"><tr><td><table class="non-framed"><tr><td><div>0</div></td><td><div class="align-right"><div>1</div></div></td></tr></table></td></tr><tr><td><div class="center"><div>x</div></div></td></tr></table></td></tr></table></div></div>

